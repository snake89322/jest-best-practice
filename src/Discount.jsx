import React, { Component, Fragment } from 'react'
import PropTypes from 'prop-types'

export class Discount extends Component {
  render () {
    const {
      isNew,
      isVip,
      discount
    } = this.props

    return (
      <Fragment>
        {
          isNew && <h1 className='newDiscount'>newDiscount</h1>
        }
        {
          isVip && <h1 className='vipDiscount'>vipDiscount</h1>
        }
        {
          <span className='discount'>{ discount }</span>
        }
      </Fragment>
    )
  }
}

Discount.propTypes = {
  isNew: PropTypes.bool,
  isVip: PropTypes.bool,
  discount: PropTypes.string
}
