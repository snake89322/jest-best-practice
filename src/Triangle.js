export class Triangle {
  constructor (a, b, c) {
    if (
      arguments.length >= 4 ||
      !a || !b || !c ||
      a < 0 || b < 0 || c < 0 ||
      !Number.isInteger(a) ||
      !Number.isInteger(b) ||
      !Number.isInteger(c)
    ) {
      this.type = 'Error'
      return this
    }

    if (a + b <= c ||
      a + c <= b ||
      b + c <= a
    ) {
      this.type = 'R1'
      return this
    }

    if (
      a + b > c ||
      a + c > b ||
      b + c > a
    ) {
      this.type = 'R2'
    }

    if (
      (a === b && b !== c) ||
      (a === c && c !== b) ||
      (b === c && c !== a)
    ) {
      this.type = 'R3'
    }

    if (a === b && b === c) {
      this.type = 'R4'
    }
  }
}
