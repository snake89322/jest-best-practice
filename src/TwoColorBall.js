export class TwoColorBall {
  constructor (redRange, blueRange) {
    this.redRange = redRange || 33
    this.blueRange = blueRange || 16
  }

  static randomRange (down, up) {
    return down + Math.ceil(Math.random() * (up - down))
  }

  generateRedNums () {
    const arr = []
    for (let i = 0; i <= this.redRange; i++) {
      arr.push(i)
    }
    return arr
  }

  removeRedNum (arr, num) {
    const index = arr.indexOf(num)
    if (index > -1) {
      arr.splice(index, 1)
    }
    return arr
  }

  roll () {
    let redNums = this.generateRedNums()
    const rollRedNums = []
    const blueNum = TwoColorBall.randomRange(0, this.blueRange)

    for (let i = 0; i <= 6; i++) {
      const rollNum = redNums[TwoColorBall.randomRange(0, redNums.length)]
      rollRedNums.push(rollNum)
      this.removeRedNum(redNums, rollNum)
    }

    return rollRedNums.sort((a, b) => (a - b)).concat(blueNum).join('-')
  }
}
