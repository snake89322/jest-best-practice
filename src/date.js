import dayjs from 'dayjs'

export function date (dateString) {
  return dayjs(dateString).format('{YYYY} MM-DDTHH:mm:ss SSS [Z] A')
}
