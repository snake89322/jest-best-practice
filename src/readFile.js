import fs from 'fs'

export function readFile (file) {
  return new Promise((resolve, reject) => {
    fs.readFile(file, 'utf-8', (error, data) => {
      if (error) reject(error)

      resolve(data)
    })
  })
}
