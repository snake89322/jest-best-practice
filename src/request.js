import fetch from 'node-fetch'

export function request (url, callback) {
  return new Promise((resolve, reject) => {
    fetch(url)
      .then(function (response) {
        if (response && response.status === 200) {
          callback && callback()
          resolve(response.json())
        } else {
          reject(new Error('request error'))
        }
      })
      .catch(function (error) {
        reject(error)
      })
  })
}
