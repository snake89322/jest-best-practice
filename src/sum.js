// 1 ≤ x1 ≤ 5
// 2 ≤ x2 ≤ 10
export function sum (x1, x2) {
  return x1 + x2
}
