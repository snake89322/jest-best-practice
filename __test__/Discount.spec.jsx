import React from 'react'
import { Discount } from '../src/Discount'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'

Enzyme.configure({ adapter: new Adapter() })

describe('discount function', () => {
  it('Rule 1', async () => {
    const props = {
      isNew: true,
      isVip: true,
      discount: '60'
    }
    const wrapper = shallow(<Discount {...props}/>)

    expect(wrapper.find('.newDiscount')).toHaveLength(1)
    expect(wrapper.find('.vipDiscount')).toHaveLength(1)
    expect(wrapper.find('.discount').text()).toEqual('60')
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('Rule 2', async () => {
    const props = {
      isNew: true,
      isVip: false,
      discount: '40'
    }
    const wrapper = shallow(<Discount {...props}/>)

    expect(wrapper.find('.newDiscount')).toHaveLength(1)
    expect(wrapper.find('.vipDiscount')).toHaveLength(0)
    expect(wrapper.find('.discount').text()).toEqual('40')
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('Rule 3', async () => {
    const props = {
      isNew: false,
      isVip: true,
      discount: '20'
    }
    const wrapper = shallow(<Discount {...props}/>)

    expect(wrapper.find('.newDiscount')).toHaveLength(0)
    expect(wrapper.find('.vipDiscount')).toHaveLength(1)
    expect(wrapper.find('.discount').text()).toEqual('20')
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('Rule 4', async () => {
    const props = {
      isNew: false,
      isVip: false,
      discount: '0'
    }
    const wrapper = shallow(<Discount {...props}/>)

    expect(wrapper.find('.newDiscount')).toHaveLength(0)
    expect(wrapper.find('.vipDiscount')).toHaveLength(0)
    expect(wrapper.find('.discount').text()).toEqual('0')
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
