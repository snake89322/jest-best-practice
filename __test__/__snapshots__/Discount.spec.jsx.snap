// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`discount function Rule 1 1`] = `
<Fragment>
  <h1
    className="newDiscount"
  >
    newDiscount
  </h1>
  <h1
    className="vipDiscount"
  >
    vipDiscount
  </h1>
  <span
    className="discount"
  >
    60
  </span>
</Fragment>
`;

exports[`discount function Rule 2 1`] = `
<Fragment>
  <h1
    className="newDiscount"
  >
    newDiscount
  </h1>
  <span
    className="discount"
  >
    40
  </span>
</Fragment>
`;

exports[`discount function Rule 3 1`] = `
<Fragment>
  <h1
    className="vipDiscount"
  >
    vipDiscount
  </h1>
  <span
    className="discount"
  >
    20
  </span>
</Fragment>
`;

exports[`discount function Rule 4 1`] = `
<Fragment>
  <span
    className="discount"
  >
    0
  </span>
</Fragment>
`;
